keps/sig-apps/20190318-Graduate-CronJob-to-Stable.md

```diff
@@ -0,0 +1,560 @@
+  - [Add Counters](#add-counters)
+  - [Add .status.conditions](#add-statusconditions)
+  - [Support Jitter for cronjobs](#support-jitter-for-cronjobs)
+  - [Support Timezone for cronjobs](#support-timezone-for-cronjobs)
+  - [CronJob umbrella issue](#cronjob-umbrella-issue)
+<!-- /toc -->
+
+## Summary
+
+[CronJob](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/) is a Kubernetes API that creates Job object on a schedule specified by a cron spec. It is in beta status since v1.8 and currently it is marked deprecated with target release for removal being v1.22. This document lays out the plan to promote it to stable.
+
+## Motivation
+
+CronJob is useful to run periodic tasks using cron like facility in a kubernetes cluster. Its `.spec` has been stable for the last few releases. We feel the API with additional `.status` information is ready to be promoted to Stable and be supported long-term by the community.
+
+## Goals
+
+- Write a new controller which will
+  - Use informers instead of polling
+  - Address some open issues with the current controller
+  - Add metrics exposing controller throughput, latency etc.
+- Extend CronJob status field
+  - lastSuccessfulTime: tracks the last time the job completed successfully
+  - nextRunTime: tracks the next time the job will be scheduled
+
+## Proposal
+
+### Plan for promoting CronJob to GA
+
+To promote to GA we would create `batch/v1/CronJob` in the [batch/v1 API](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/batch/v1/types.go).
+
+We shall have dual implementation of controller (old and new) co-exist.
+Through a feature flag the new controller implementation is controlled.
+- alpha (disabled, can be enabled)
+- beta (enabled, can be disabled)
+- GA (enabled, cannot be disabled)
+
+API will track the controller since GA-ing the API implies GA-ing the scalable implementation of the API.
+
+### Existing controller
+
+The current implementation of the CronJob controller is different that the other workload controllers. GA workload controllers use informers and caches to reduce the load on API server. Whereas the cronjob controller does a periodic poll and sweep of all the objects and acts on them. The CronJob controller has only one worker doing this.
+
+1. syncs all CronJob objects [every 10 seconds](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller.go#L98).
+2. Using pager library, gets all Pods and all CronJobs and [processes them one by one](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/cronjob/cronjob_controller.go#L136)
+
+This is not a scalable design and ends up loading the API server. Also this does not follow the [recommended guidelines](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-api-machinery/controllers.md) for building controllers.
+
+### New controller
+
+With this approach we aim to:
```

Comment:
nit: scale issues already include 2 and 3
```suggestion
With this approach we aim to reduce the potential scale issues (e.g. load on API server and memory usage) when using lots of CronJob objects.
```

Reply:
Done.